torch/distributed/_tensor/experimental/attention.py

```diff
@@ -634,27 +724,136 @@ def backward_hook(grad: torch.Tensor) -> None:
+            return new_outputs[0]
+
+        return tuple(new_outputs)
+
+    # TODO: provide a more robust way to replace SDPA.
+    # Currently we use monkey patch to replace scaled_dot_product_attention with the
+    # wrapped fn. This is okay if users do `import torch.nn.functional` but will not
+    # work if users do `import torch.nn.functional.scaled_dot_product_attention`.
+    _distribute_function(
+        F.scaled_dot_product_attention,
+        F,
+        mesh,
+        attention_input_fn,
+        attention_output_fn,
+    )
+
+    with _enable_cp_dispatcher():
+        yield
+
+    _restore_function(F.scaled_dot_product_attention, F)
+
+
+def _get_sequence_shard(
+    buffer: torch.Tensor, mesh: DeviceMesh, seq_dim: int
+) -> torch.Tensor:
+    return buffer.chunk(mesh.size(), dim=seq_dim)[mesh.get_local_rank()]
+
+
+def _context_parallel_buffers(
+    mesh: DeviceMesh,
+    buffers: List[torch.Tensor],
+    buffer_seq_dims: List[int],
+) -> List[torch.Tensor]:
+    """Shard the buffers along the sequence dimensions according to CP rules."""
+    new_buffers = []
+    for buffer, seq_dim in zip(buffers, buffer_seq_dims):
+        new_buffers.append(_get_sequence_shard(buffer, mesh, seq_dim))
+
+    return new_buffers
+
+
+@contextlib.contextmanager
+@torch.no_grad()
+def context_parallel(
+    mesh: DeviceMesh,
+    *,
+    buffers: Optional[List[torch.Tensor]] = None,
+    buffer_seq_dims: Optional[List[int]] = None,
+    no_restore_buffers: Optional[Set[torch.Tensor]] = None,
+) -> Generator[None, None, None]:
     """
-    Calculate is_causal behavior for each KV block. The attention can either be
-    calculated in full, not at all or with the causal mask applied.
+
+    ``context_parallel`` is an experimental API to enable context
+    parallelism (CP). This API performs two actions: 1) patch the SDPA
+    (``torch.nn.functional.scaled_dot_product_attention``) with the CP-enabled
+    one, 2) shard ``buffers`` along the sequence dimension and each rank will
+    preserve the corresponding shard according ``mesh``.
+
+    Args:
+        mesh (:class:`DeviceMesh`): the device mesh for the context parallelism.
+        buffers (Optional[List[torch.Tensor]]): buffers that the usage depend
+            on the sequence dimension. Examples are input batch, labels and
+            positional embedding buffers. These buffers must be sharded along
+            the sequence dimension to ensure the accuracy. The sharding will
+            happen in-place, the buffer's shape will change within the context.
+            The buffers will be restored after the context finishes.
+            ``no_restore_buffers`` can be used to specify which buffers don't
+            need to be restored. Note that ``buffers`` should not contain any
+            nn.Parameter.
+        buffer_seq_dims (Optional[List[int]]): the sequence dimensions of ``buffers``.
+        no_restore_buffers (Optional[Set[torch.Tensor]]): buffers in these set
```

Comment:
When user would determine whether the buffers should be restored or not? iiuc input batch/labels does not need to be restored? I wonder if there's a implication that this context manager need to be applied during the whole model fwd/bwd?

Reply:
ye, this is mostly for efficiency, so batch and label don't need to be restored. And yes, the context manager must be applied during the whole fwd/bwd.